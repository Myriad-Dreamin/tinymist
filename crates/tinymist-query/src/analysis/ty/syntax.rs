//! Type checking on source file

use std::collections::BTreeMap;

use typst::{
    foundations::Value,
    syntax::{
        ast::{self, AstNode},
        LinkedNode, SyntaxKind,
    },
};

use super::*;
use crate::{adt::interner::Interned, ty::*};

impl<'a, 'w> TypeChecker<'a, 'w> {
    pub(crate) fn check_syntax(&mut self, root: LinkedNode) -> Option<Ty> {
        Some(match root.kind() {
            SyntaxKind::Markup => return self.check_in_mode(root, InterpretMode::Markup),
            SyntaxKind::Math => return self.check_in_mode(root, InterpretMode::Math),
            SyntaxKind::Code => return self.check_in_mode(root, InterpretMode::Code),
            SyntaxKind::CodeBlock => return self.check_in_mode(root, InterpretMode::Code),
            SyntaxKind::ContentBlock => return self.check_in_mode(root, InterpretMode::Markup),

            // todo: space effect
            SyntaxKind::Space => Ty::Space,
            SyntaxKind::Parbreak => Ty::Space,

            SyntaxKind::Text => Ty::Content,
            SyntaxKind::Linebreak => Ty::Content,
            SyntaxKind::Escape => Ty::Content,
            SyntaxKind::Shorthand => Ty::Content,
            SyntaxKind::SmartQuote => Ty::Content,
            SyntaxKind::Raw => Ty::Content,
            SyntaxKind::RawLang => Ty::Content,
            SyntaxKind::RawDelim => Ty::Content,
            SyntaxKind::RawTrimmed => Ty::Content,
            SyntaxKind::Link => Ty::Content,
            SyntaxKind::Label => Ty::Content,
            SyntaxKind::Ref => Ty::Content,
            SyntaxKind::RefMarker => Ty::Content,
            SyntaxKind::HeadingMarker => Ty::Content,
            SyntaxKind::EnumMarker => Ty::Content,
            SyntaxKind::ListMarker => Ty::Content,
            SyntaxKind::TermMarker => Ty::Content,
            SyntaxKind::MathAlignPoint => Ty::Content,
            SyntaxKind::MathPrimes => Ty::Content,

            SyntaxKind::Strong => return self.check_children(root),
            SyntaxKind::Emph => return self.check_children(root),
            SyntaxKind::Heading => return self.check_children(root),
            SyntaxKind::ListItem => return self.check_children(root),
            SyntaxKind::EnumItem => return self.check_children(root),
            SyntaxKind::TermItem => return self.check_children(root),
            SyntaxKind::Equation => return self.check_children(root),
            SyntaxKind::MathDelimited => return self.check_children(root),
            SyntaxKind::MathAttach => return self.check_children(root),
            SyntaxKind::MathFrac => return self.check_children(root),
            SyntaxKind::MathRoot => return self.check_children(root),

            SyntaxKind::LoopBreak => Ty::None,
            SyntaxKind::LoopContinue => Ty::None,
            SyntaxKind::FuncReturn => Ty::None,
            SyntaxKind::Error => Ty::None,
            SyntaxKind::Eof => Ty::None,

            SyntaxKind::None => Ty::None,
            SyntaxKind::Auto => Ty::Auto,
            SyntaxKind::Break => Ty::FlowNone,
            SyntaxKind::Continue => Ty::FlowNone,
            SyntaxKind::Return => Ty::FlowNone,
            SyntaxKind::Ident => return self.check_ident(root, InterpretMode::Code),
            SyntaxKind::MathIdent => return self.check_ident(root, InterpretMode::Math),
            SyntaxKind::Bool
            | SyntaxKind::Int
            | SyntaxKind::Float
            | SyntaxKind::Numeric
            | SyntaxKind::Str => {
                return self
                    .ctx
                    .mini_eval(root.cast()?)
                    .map(|v| (Ty::Value(InsTy::new(v))))
            }
            SyntaxKind::Parenthesized => return self.check_children(root),
            SyntaxKind::Array => return self.check_array(root),
            SyntaxKind::Dict => return self.check_dict(root),
            SyntaxKind::Unary => return self.check_unary(root),
            SyntaxKind::Binary => return self.check_binary(root),
            SyntaxKind::FieldAccess => return self.check_field_access(root),
            SyntaxKind::FuncCall => return self.check_func_call(root),
            SyntaxKind::Args => return self.check_args(root),
            SyntaxKind::Closure => return self.check_closure(root),
            SyntaxKind::LetBinding => return self.check_let(root),
            SyntaxKind::SetRule => return self.check_set(root),
            SyntaxKind::ShowRule => return self.check_show(root),
            SyntaxKind::Contextual => return self.check_contextual(root),
            SyntaxKind::Conditional => return self.check_conditional(root),
            SyntaxKind::WhileLoop => return self.check_while_loop(root),
            SyntaxKind::ForLoop => return self.check_for_loop(root),
            SyntaxKind::ModuleImport => return self.check_module_import(root),
            SyntaxKind::ModuleInclude => return self.check_module_include(root),
            SyntaxKind::Destructuring => return self.check_destructuring(root),
            SyntaxKind::DestructAssignment => return self.check_destruct_assign(root),

            // Rest all are clauses
            SyntaxKind::LineComment => Ty::Clause,
            SyntaxKind::BlockComment => Ty::Clause,
            SyntaxKind::Named => Ty::Clause,
            SyntaxKind::Keyed => Ty::Clause,
            SyntaxKind::Spread => Ty::Clause,
            SyntaxKind::Params => Ty::Clause,
            SyntaxKind::ImportItems => Ty::Clause,
            SyntaxKind::RenamedImportItem => Ty::Clause,
            SyntaxKind::Hash => Ty::Clause,
            SyntaxKind::LeftBrace => Ty::Clause,
            SyntaxKind::RightBrace => Ty::Clause,
            SyntaxKind::LeftBracket => Ty::Clause,
            SyntaxKind::RightBracket => Ty::Clause,
            SyntaxKind::LeftParen => Ty::Clause,
            SyntaxKind::RightParen => Ty::Clause,
            SyntaxKind::Comma => Ty::Clause,
            SyntaxKind::Semicolon => Ty::Clause,
            SyntaxKind::Colon => Ty::Clause,
            SyntaxKind::Star => Ty::Clause,
            SyntaxKind::Underscore => Ty::Clause,
            SyntaxKind::Dollar => Ty::Clause,
            SyntaxKind::Plus => Ty::Clause,
            SyntaxKind::Minus => Ty::Clause,
            SyntaxKind::Slash => Ty::Clause,
            SyntaxKind::Hat => Ty::Clause,
            SyntaxKind::Prime => Ty::Clause,
            SyntaxKind::Dot => Ty::Clause,
            SyntaxKind::Eq => Ty::Clause,
            SyntaxKind::EqEq => Ty::Clause,
            SyntaxKind::ExclEq => Ty::Clause,
            SyntaxKind::Lt => Ty::Clause,
            SyntaxKind::LtEq => Ty::Clause,
            SyntaxKind::Gt => Ty::Clause,
            SyntaxKind::GtEq => Ty::Clause,
            SyntaxKind::PlusEq => Ty::Clause,
            SyntaxKind::HyphEq => Ty::Clause,
            SyntaxKind::StarEq => Ty::Clause,
            SyntaxKind::SlashEq => Ty::Clause,
            SyntaxKind::Dots => Ty::Clause,
            SyntaxKind::Arrow => Ty::Clause,
            SyntaxKind::Root => Ty::Clause,
            SyntaxKind::Not => Ty::Clause,
            SyntaxKind::And => Ty::Clause,
            SyntaxKind::Or => Ty::Clause,
            SyntaxKind::Let => Ty::Clause,
            SyntaxKind::Set => Ty::Clause,
            SyntaxKind::Show => Ty::Clause,
            SyntaxKind::Context => Ty::Clause,
            SyntaxKind::If => Ty::Clause,
            SyntaxKind::Else => Ty::Clause,
            SyntaxKind::For => Ty::Clause,
            SyntaxKind::In => Ty::Clause,
            SyntaxKind::While => Ty::Clause,
            SyntaxKind::Import => Ty::Clause,
            SyntaxKind::Include => Ty::Clause,
            SyntaxKind::As => Ty::Clause,
        })
    }

    fn check_in_mode(&mut self, root: LinkedNode, into_mode: InterpretMode) -> Option<Ty> {
        let mode = self.mode;
        self.mode = into_mode;
        let res = self.check_children(root);
        self.mode = mode;
        res
    }

    fn check_children(&mut self, root: LinkedNode<'_>) -> Option<Ty> {
        let mut joiner = Joiner::default();

        for child in root.children() {
            joiner.join(self.check(child));
        }
        Some(joiner.finalize())
    }

    fn check_ident(&mut self, root: LinkedNode<'_>, mode: InterpretMode) -> Option<Ty> {
        let ident: ast::Ident = root.cast()?;
        let ident_ref = IdentRef {
            name: ident.get().to_string(),
            range: root.range(),
        };

        self.get_var(root.span(), ident_ref).or_else(|| {
            let s = root.span();
            let v = resolve_global_value(self.ctx, root, mode == InterpretMode::Math)?;
            Some(Ty::Value(InsTy::new_at(v, s)))
        })
    }

    fn check_array(&mut self, root: LinkedNode<'_>) -> Option<Ty> {
        let _arr: ast::Array = root.cast()?;

        let mut elements = Vec::new();

        for elem in root.children() {
            let ty = self.check(elem);
            if matches!(ty, Ty::Clause | Ty::Space) {
                continue;
            }
            elements.push(ty);
        }

        Some(Ty::Tuple(elements.into()))
    }

    fn check_dict(&mut self, root: LinkedNode<'_>) -> Option<Ty> {
        let dict: ast::Dict = root.cast()?;

        let mut fields = Vec::new();

        for field in dict.items() {
            match field {
                ast::DictItem::Named(n) => {
                    let name = n.name().into();
                    let value = self.check_expr_in(n.expr().span(), root.clone());
                    fields.push((name, value, n.span()));
                }
                ast::DictItem::Keyed(k) => {
                    let key = self.ctx.const_eval(k.key());
                    if let Some(Value::Str(key)) = key {
                        let value = self.check_expr_in(k.expr().span(), root.clone());
                        fields.push((key.into(), value, k.span()));
                    }
                }
                // todo: var dict union
                ast::DictItem::Spread(_s) => {}
            }
        }

        Some(Ty::Dict(RecordTy::new(fields)))
    }

    fn check_unary(&mut self, root: LinkedNode<'_>) -> Option<Ty> {
        let unary: ast::Unary = root.cast()?;

        if let Some(constant) = self.ctx.mini_eval(ast::Expr::Unary(unary)) {
            return Some(Ty::Value(InsTy::new(constant)));
        }

        let op = unary.op();

        let lhs = self.check_expr_in(unary.expr().span(), root).into();
        let op = match op {
            ast::UnOp::Pos => UnaryOp::Pos,
            ast::UnOp::Neg => UnaryOp::Neg,
            ast::UnOp::Not => UnaryOp::Not,
        };

        Some(Ty::Unary(TypeUnary::new(op, lhs)))
    }

    fn check_binary(&mut self, root: LinkedNode<'_>) -> Option<Ty> {
        let binary: ast::Binary = root.cast()?;

        if let Some(constant) = self.ctx.mini_eval(ast::Expr::Binary(binary)) {
            return Some(Ty::Value(InsTy::new(constant)));
        }

        let op = binary.op();
        let lhs_span = binary.lhs().span();
        let lhs = self.check_expr_in(lhs_span, root.clone());
        let rhs_span = binary.rhs().span();
        let rhs = self.check_expr_in(rhs_span, root);

        match op {
            ast::BinOp::Add | ast::BinOp::Sub | ast::BinOp::Mul | ast::BinOp::Div => {}
            ast::BinOp::Eq | ast::BinOp::Neq | ast::BinOp::Leq | ast::BinOp::Geq => {
                self.check_comparable(&lhs, &rhs);
                self.possible_ever_be(&lhs, &rhs);
                self.possible_ever_be(&rhs, &lhs);
            }
            ast::BinOp::Lt | ast::BinOp::Gt => {
                self.check_comparable(&lhs, &rhs);
            }
            ast::BinOp::And | ast::BinOp::Or => {
                self.constrain(&lhs, &Ty::Boolean(None));
                self.constrain(&rhs, &Ty::Boolean(None));
            }
            ast::BinOp::NotIn | ast::BinOp::In => {
                self.check_containing(&rhs, &lhs, op == ast::BinOp::In);
            }
            ast::BinOp::Assign => {
                self.check_assignable(&lhs, &rhs);
                self.possible_ever_be(&lhs, &rhs);
            }
            ast::BinOp::AddAssign
            | ast::BinOp::SubAssign
            | ast::BinOp::MulAssign
            | ast::BinOp::DivAssign => {
                self.check_assignable(&lhs, &rhs);
            }
        }

        let res = Ty::Binary(TypeBinary::new(op, lhs.into(), rhs.into()));

        Some(res)
    }

    fn check_field_access(&mut self, root: LinkedNode<'_>) -> Option<Ty> {
        let field_access: ast::FieldAccess = root.cast()?;

        let ty = self.check_expr_in(field_access.target().span(), root.clone());
        let field = field_access.field().get().clone();

        Some(Ty::Select(SelectTy::new(ty.into(), field.into())))
    }

    fn check_func_call(&mut self, root: LinkedNode<'_>) -> Option<Ty> {
        let func_call: ast::FuncCall = root.cast()?;

        let args = self.check_expr_in(func_call.args().span(), root.clone());
        let callee = self.check_expr_in(func_call.callee().span(), root.clone());

        log::debug!("func_call: {callee:?} with {args:?}");

        if let Ty::Args(args) = args {
            let mut worker = ApplyTypeChecker {
                base: self,
                call_site: func_call.callee().span(),
                args: func_call.args(),
                resultant: vec![],
            };
            callee.call(&args, true, &mut worker);
            return Some(Ty::from_types(worker.resultant.into_iter()));
        }

        None
    }

    fn check_args(&mut self, root: LinkedNode<'_>) -> Option<Ty> {
        let args: ast::Args = root.cast()?;

        let mut args_res = Vec::new();
        let mut named = vec![];

        for arg in args.items() {
            match arg {
                ast::Arg::Pos(e) => {
                    args_res.push(self.check_expr_in(e.span(), root.clone()));
                }
                ast::Arg::Named(n) => {
                    let value = self.check_expr_in(n.expr().span(), root.clone());
                    named.push((n.name().into(), value));
                }
                // todo
                ast::Arg::Spread(_w) => {}
            }
        }

        let args = ArgsTy::new(args_res.into_iter(), named.into_iter(), None, None);

        Some(Ty::Args(args.into()))
    }

    fn check_closure(&mut self, root: LinkedNode<'_>) -> Option<Ty> {
        let closure: ast::Closure = root.cast()?;

        // let _params = self.check_expr_in(closure.params().span(), root.clone());

        let mut pos = vec![];
        let mut named = BTreeMap::new();
        let mut rest = None;

        for param in closure.params().children() {
            match param {
                ast::Param::Pos(pattern) => {
                    pos.push(self.check_pattern(pattern, Ty::Any, root.clone()));
                }
                ast::Param::Named(e) => {
                    let exp = self.check_expr_in(e.expr().span(), root.clone());
                    let v = self.get_var(e.name().span(), to_ident_ref(&root, e.name())?)?;
                    // todo: this is less efficient than v.lbs.push(exp), we may have some idea to
                    // optimize it, so I put a todo here.
                    self.constrain(&exp, &v);
                    named.insert(e.name().into(), v);
                }
                // todo: spread left/right
                ast::Param::Spread(a) => {
                    if let Some(e) = a.sink_ident() {
                        let exp = Ty::Builtin(BuiltinTy::Args);
                        let v = self.get_var(e.span(), to_ident_ref(&root, e)?)?;
                        self.constrain(&exp, &v);
                        rest = Some(v);
                    }
                    // todo: ..(args)
                }
            }
        }

        let body = self.check_expr_in(closure.body().span(), root);

        let named: Vec<(Interned<str>, Ty)> = named.into_iter().collect();

        // freeze the signature
        for pos in pos.iter() {
            self.weaken(pos);
        }
        for (_, named) in named.iter() {
            self.weaken(named);
        }
        if let Some(rest) = &rest {
            self.weaken(rest);
        }

        let sig = SigTy::new(pos.into_iter(), named.into_iter(), rest, Some(body));
        Some(Ty::Func(sig.into()))
    }

    fn check_let(&mut self, root: LinkedNode<'_>) -> Option<Ty> {
        let let_binding: ast::LetBinding = root.cast()?;

        match let_binding.kind() {
            ast::LetBindingKind::Closure(c) => {
                // let _name = let_binding.name().get().to_string();
                let value = let_binding
                    .init()
                    .map(|init| self.check_expr_in(init.span(), root.clone()))
                    .unwrap_or_else(|| Ty::Infer);

                let v = self.get_var(c.span(), to_ident_ref(&root, c)?)?;
                self.constrain(&value, &v);
                // todo lbs is the lexical signature.
            }
            ast::LetBindingKind::Normal(pattern) => {
                // let _name = let_binding.name().get().to_string();
                let value = let_binding
                    .init()
                    .map(|init| self.check_expr_in(init.span(), root.clone()))
                    .unwrap_or_else(|| Ty::Infer);

                self.check_pattern(pattern, value, root.clone());
            }
        }

        Some(Ty::Any)
    }

    // todo: merge with func call, and regard difference (may be here)
    fn check_set(&mut self, root: LinkedNode<'_>) -> Option<Ty> {
        let set_rule: ast::SetRule = root.cast()?;

        let callee = self.check_expr_in(set_rule.target().span(), root.clone());
        let args = self.check_expr_in(set_rule.args().span(), root.clone());
        let _cond = set_rule
            .condition()
            .map(|cond| self.check_expr_in(cond.span(), root.clone()));

        log::debug!("set rule: {callee:?} with {args:?}");

        if let Ty::Args(args) = args {
            let mut worker = ApplyTypeChecker {
                base: self,
                call_site: set_rule.target().span(),
                args: set_rule.args(),
                resultant: vec![],
            };
            callee.call(&args, true, &mut worker);
            return Some(Ty::from_types(worker.resultant.into_iter()));
        }

        None
    }

    fn check_show(&mut self, root: LinkedNode<'_>) -> Option<Ty> {
        let show_rule: ast::ShowRule = root.cast()?;

        let _selector = show_rule
            .selector()
            .map(|sel| self.check_expr_in(sel.span(), root.clone()));
        let t = show_rule.transform();
        // todo: infer it type by selector
        let _transform = self.check_expr_in(t.span(), root.clone());

        Some(Ty::Any)
    }

    // currently we do nothing on contextual
    fn check_contextual(&mut self, root: LinkedNode<'_>) -> Option<Ty> {
        let contextual: ast::Contextual = root.cast()?;

        let body = self.check_expr_in(contextual.body().span(), root);

        Some(Ty::Unary(TypeUnary::new(UnaryOp::Context, body.into())))
    }

    fn check_conditional(&mut self, root: LinkedNode<'_>) -> Option<Ty> {
        let conditional: ast::Conditional = root.cast()?;

        let cond = self.check_expr_in(conditional.condition().span(), root.clone());
        let then = self.check_expr_in(conditional.if_body().span(), root.clone());
        let else_ = conditional
            .else_body()
            .map(|else_body| self.check_expr_in(else_body.span(), root.clone()))
            .unwrap_or(Ty::None);

        Some(Ty::If(IfTy::new(cond.into(), then.into(), else_.into())))
    }

    fn check_while_loop(&mut self, root: LinkedNode<'_>) -> Option<Ty> {
        let while_loop: ast::WhileLoop = root.cast()?;

        let _cond = self.check_expr_in(while_loop.condition().span(), root.clone());
        let _body = self.check_expr_in(while_loop.body().span(), root);

        Some(Ty::Any)
    }

    fn check_for_loop(&mut self, root: LinkedNode<'_>) -> Option<Ty> {
        let for_loop: ast::ForLoop = root.cast()?;

        let _iter = self.check_expr_in(for_loop.iterable().span(), root.clone());
        let _pattern = self.check_expr_in(for_loop.pattern().span(), root.clone());
        let _body = self.check_expr_in(for_loop.body().span(), root);

        Some(Ty::Any)
    }

    fn check_module_import(&mut self, root: LinkedNode<'_>) -> Option<Ty> {
        let _module_import: ast::ModuleImport = root.cast()?;

        // check all import items

        Some(Ty::None)
    }

    fn check_module_include(&mut self, _root: LinkedNode<'_>) -> Option<Ty> {
        Some(Ty::Content)
    }

    fn check_destructuring(&mut self, _root: LinkedNode<'_>) -> Option<Ty> {
        Some(Ty::Any)
    }

    fn check_destruct_assign(&mut self, _root: LinkedNode<'_>) -> Option<Ty> {
        Some(Ty::None)
    }
    fn check_expr_in(&mut self, span: Span, root: LinkedNode<'_>) -> Ty {
        root.find(span)
            .map(|node| self.check(node))
            .unwrap_or(Ty::Undef)
    }

    fn check_pattern(&mut self, pattern: ast::Pattern<'_>, value: Ty, root: LinkedNode<'_>) -> Ty {
        self.check_pattern_(pattern, value, root)
            .unwrap_or(Ty::Undef)
    }

    fn check_pattern_(
        &mut self,
        pattern: ast::Pattern<'_>,
        value: Ty,
        root: LinkedNode<'_>,
    ) -> Option<Ty> {
        Some(match pattern {
            ast::Pattern::Normal(ast::Expr::Ident(ident)) => {
                let v = self.get_var(ident.span(), to_ident_ref(&root, ident)?)?;
                self.constrain(&value, &v);
                v
            }
            ast::Pattern::Normal(_) => Ty::Any,
            ast::Pattern::Placeholder(_) => Ty::Any,
            ast::Pattern::Parenthesized(exp) => self.check_pattern(exp.pattern(), value, root),
            // todo: pattern
            ast::Pattern::Destructuring(_destruct) => Ty::Any,
        })
    }
}
