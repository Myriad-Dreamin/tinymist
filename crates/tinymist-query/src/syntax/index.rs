use std::str::FromStr;

use reflexo_typst::package::PackageSpec;
use rustc_hash::FxHashSet;

use crate::{adt::interner::Interned, prelude::*};

#[derive(Default)]
pub struct IndexInfo {
    pub(crate) paths: FxHashSet<Interned<str>>,
    pub(crate) packages: FxHashSet<PackageSpec>,
    pub(crate) identifiers: FxHashSet<Interned<str>>,
}

#[comemo::memoize]
pub fn get_index_info(src: &Source) -> Arc<IndexInfo> {
    let root = src.root();
    let mut worker = IndexWorker {
        info: IndexInfo::default(),
    };
    worker.visit(root);
    Arc::new(worker.info)
}

struct IndexWorker {
    info: IndexInfo,
}

impl IndexWorker {
    fn visit(&mut self, node: &SyntaxNode) {
        match node.cast::<ast::Expr>() {
            Some(ast::Expr::Str(path_str)) => {
                if path_str.to_untyped().text().len() > 65536 {
                    // skip long strings
                    return;
                }
                let path_str = path_str.get();

                if path_str.starts_with('@') {
                    let pkg_spec = PackageSpec::from_str(&path_str).ok();
                    if let Some(pkg_spec) = pkg_spec {
                        self.info.identifiers.insert(pkg_spec.name.clone().into());
                        self.info.packages.insert(pkg_spec);
                    }
                    return;
                }
                let path = Path::new(path_str.as_str());
                let name = path.file_name().unwrap_or_default().to_str();
                if let Some(name) = name {
                    self.info.paths.insert(name.into());
                }
            }
            Some(ast::Expr::MathIdent(ident)) => {
                self.info.identifiers.insert(ident.get().into());
            }
            Some(ast::Expr::Ident(ident)) => {
                self.info.identifiers.insert(ident.get().into());
            }
            _ => {}
        }

        for child in node.children() {
            self.visit(child);
        }
    }
}
