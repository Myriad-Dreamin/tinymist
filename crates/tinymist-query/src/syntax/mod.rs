//! Analyzing the syntax of a source file.
//!
//! This module must hide all **AST details** from the rest of the codebase.

// todo: remove this
#![allow(missing_docs)]

use ecow::EcoString;
pub use tinymist_analysis::import::*;
pub(crate) mod lexical_hierarchy;
pub use lexical_hierarchy::*;
pub(crate) mod matcher;
pub use matcher::*;
pub(crate) mod module;
pub use module::*;
pub(crate) mod comment;
pub use comment::*;
pub(crate) mod expr;
pub use expr::*;
pub(crate) mod docs;
pub use docs::*;

use core::fmt;
use std::ops::Range;

use serde::{Deserialize, Serialize};

/// A flat and transient reference to some symbol in a source file.
///
/// It is transient because it is not guaranteed to be valid after the source
/// file is modified.
#[derive(Debug, Clone, Hash, PartialEq, Eq)]
pub struct IdentRef {
    /// The name of the symbol.
    pub name: EcoString,
    /// The byte range of the symbol in the source file.
    pub range: Range<usize>,
}

impl PartialOrd for IdentRef {
    fn partial_cmp(&self, other: &Self) -> Option<std::cmp::Ordering> {
        Some(self.cmp(other))
    }
}

impl Ord for IdentRef {
    fn cmp(&self, other: &Self) -> std::cmp::Ordering {
        self.name
            .cmp(&other.name)
            .then_with(|| self.range.start.cmp(&other.range.start))
    }
}

impl fmt::Display for IdentRef {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "{}@{:?}", self.name, self.range)
    }
}

impl Serialize for IdentRef {
    fn serialize<S: serde::Serializer>(&self, serializer: S) -> Result<S::Ok, S::Error> {
        let s = self.to_string();
        serializer.serialize_str(&s)
    }
}

impl<'de> Deserialize<'de> for IdentRef {
    fn deserialize<D: serde::Deserializer<'de>>(deserializer: D) -> Result<Self, D::Error> {
        let s = String::deserialize(deserializer)?;
        let (name, range) = {
            let mut parts = s.split('@');
            let name = parts.next().ok_or_else(|| {
                serde::de::Error::custom("expected name@range, but found empty string")
            })?;
            let range = parts.next().ok_or_else(|| {
                serde::de::Error::custom("expected name@range, but found no range")
            })?;
            // let range = range
            //     .parse()
            //     .map_err(|e| serde::de::Error::custom(format!("failed to parse range:
            // {}", e)))?;
            let st_ed = range
                .split("..")
                .map(|s| {
                    s.parse().map_err(|e| {
                        serde::de::Error::custom(format!("failed to parse range: {e}"))
                    })
                })
                .collect::<Result<Vec<usize>, _>>()?;
            if st_ed.len() != 2 {
                return Err(serde::de::Error::custom("expected range to have 2 parts"));
            }
            (name, st_ed[0]..st_ed[1])
        };
        Ok(IdentRef {
            name: name.into(),
            range,
        })
    }
}

/// A flat and transient reference to some symbol in a source file.
///
/// See [`IdentRef`] for definition of a "transient" reference.
#[derive(Debug, Clone, Serialize)]
pub struct IdentDef {
    /// The name of the symbol.
    pub name: EcoString,
    /// The kind of the symbol.
    pub kind: LexicalKind,
    /// The byte range of the symbol in the source file.
    pub range: Range<usize>,
}
