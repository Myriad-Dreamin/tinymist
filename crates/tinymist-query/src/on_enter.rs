//! <https://github.com/rust-lang/rust-analyzer/blob/master/docs/dev/lsp-extensions.md#on-enter>

use typst_shim::syntax::LinkedNodeExt;

use crate::{prelude::*, syntax::node_ancestors, SyntaxRequest};

/// The [`experimental/onEnter`] request is sent from client to server to handle
/// the <kbd>Enter</kbd> key press.
///
/// - `kbd:Enter` inside triple-slash comments automatically inserts `///`
/// - `kbd:Enter` in the middle or after a trailing space in `//` inserts `//`
/// - `kbd:Enter` inside `//!` doc comments automatically inserts `//!`
///
/// [`experimental/onEnter`]: https://github.com/rust-lang/rust-analyzer/blob/master/docs/dev/lsp-extensions.md#on-enter
///
/// # Compatibility
///
/// This request was introduced in specification version 3.10.0.
#[derive(Debug, Clone)]
pub struct OnEnterRequest {
    /// The path of the document to get folding ranges for.
    pub path: PathBuf,
    /// The source code position to request for.
    pub position: LspPosition,
}

impl SyntaxRequest for OnEnterRequest {
    type Response = Vec<TextEdit>;

    fn request(
        self,
        source: &Source,
        position_encoding: PositionEncoding,
    ) -> Option<Self::Response> {
        let root = LinkedNode::new(source.root());
        let cursor = lsp_to_typst::position(self.position, position_encoding, source)?;
        let leaf = root.leaf_at_compat(cursor)?;

        let worker = OnEnterWorker {
            source,
            position_encoding,
        };

        if matches!(leaf.kind(), SyntaxKind::LineComment) {
            return worker.enter_line_doc_comment(&leaf, cursor);
        }

        let math_node =
            node_ancestors(&leaf).find(|node| matches!(node.kind(), SyntaxKind::Equation));
        if let Some(mn) = math_node {
            return worker.enter_block_math(mn, cursor);
        }

        None
    }
}

struct OnEnterWorker<'a> {
    source: &'a Source,
    position_encoding: PositionEncoding,
}

impl OnEnterWorker<'_> {
    fn indent_of(&self, of: usize) -> String {
        let all_text = self.source.text();
        let start = all_text[..of].rfind('\n').map(|e| e + 1);
        let indent_size = all_text[start.unwrap_or_default()..of].chars().count();
        " ".repeat(indent_size)
    }

    fn enter_line_doc_comment(&self, leaf: &LinkedNode, cursor: usize) -> Option<Vec<TextEdit>> {
        let skipper = |n: &LinkedNode| {
            matches!(
                n.kind(),
                SyntaxKind::Space | SyntaxKind::Linebreak | SyntaxKind::LineComment
            )
        };
        let parent = leaf.parent()?;
        let till_curr = parent.children().take(leaf.index());
        let first_index = till_curr.rev().take_while(skipper).count();
        let comment_group_cnt = parent
            .children()
            .skip(leaf.index().saturating_sub(first_index))
            .take_while(skipper)
            .filter(|e| matches!(e.kind(), SyntaxKind::LineComment))
            .count();

        let comment_prefix = {
            let mut p = unscanny::Scanner::new(leaf.text());
            p.eat_while('/');
            p.eat_if('!');
            p.before()
        };

        // Continuing single-line non-doc comments (like this one :) ) is annoying
        if comment_group_cnt <= 1 && comment_prefix == "//" {
            return None;
        }

        let indent = self.indent_of(leaf.offset());
        // todo: remove_trailing_whitespace

        let rng = cursor..cursor;

        let edit = TextEdit {
            range: typst_to_lsp::range(rng, self.source, self.position_encoding),
            new_text: format!("\n{indent}{comment_prefix} $0"),
        };

        Some(vec![edit])
    }

    fn enter_block_math(&self, math_node: &LinkedNode<'_>, cursor: usize) -> Option<Vec<TextEdit>> {
        let o = math_node.range();
        if !o.contains(&cursor) {
            return None;
        }

        let all_text = self.source.text();
        let math_text = &all_text[o.clone()];
        let content = math_text.trim_start_matches('$').trim_end_matches('$');
        if !content.trim().is_empty() {
            return None;
        }

        let indent = self.indent_of(o.start);
        let rng = cursor..cursor;
        let edit = TextEdit {
            range: typst_to_lsp::range(rng, self.source, self.position_encoding),
            // todo: read indent configuration
            new_text: if !content.contains('\n') {
                format!("\n{indent}  $0\n{indent}")
            } else {
                format!("\n{indent}  $0")
            },
        };

        Some(vec![edit])
    }
}
