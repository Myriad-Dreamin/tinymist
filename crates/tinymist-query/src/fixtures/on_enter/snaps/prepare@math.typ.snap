---
source: crates/tinymist-query/src/on_enter.rs
description: "On Enter on  2..2 */\n$||$\n)"
expression: "JsonRepr::new_redacted(result, &REDACT_LOC)"
input_file: crates/tinymist-query/src/fixtures/on_enter/math.typ
---
[
 {
  "newText": "\n  $0\n",
  "range": "1:1:1:1"
 }
]
